File: rec_tf.py

# coding: utf-8

# # Tensor Factorization for Recommender Systems
# ### - CP factorization on tensor (user, music, context)
# ### - Written by ByungSoo Jeon, NAVER LABS

# In[59]:

# Add system path to use scikit-tensor Library
import sys
import sktensor as skt
import numpy as np
import math

# Set logging to DEBUG to see CP-ALS information
import logging
logging.basicConfig(level=logging.DEBUG)

# Use sparse matrix format
from scipy import sparse


# ### - Regularized matrix factorization

# In[60]:

def matrix_factorization(R, P, Q, K, steps=5000, alpha=0.0002, beta=0.02):
    Q = Q.T
    e = 0
    for step in range(steps):
        for i in range(len(R)):
            for j in range(len(R[i])):
                if R[i][j] > 0:
                    eij = R[i][j] - numpy.dot(P[i,:],Q[:,j])
                    for k in range(K):
                        P[i][k] = P[i][k] + alpha * (2 * eij * Q[k][j] - beta * P[i][k])
                        Q[k][j] = Q[k][j] + alpha * (2 * eij * P[i][k] - beta * Q[k][j])
        eR = numpy.dot(P,Q)
        e = 0
        for i in range(len(R)):
            for j in range(len(R[i])):
                if R[i][j] > 0:
                    e = e + pow(R[i][j] - numpy.dot(P[i,:],Q[:,j]), 2)
#                     for k in range(K):
#                         e = e + (beta/2) * ( pow(P[i][k],2) + pow(Q[k][j],2) )
        if e < 0.001:
            break
    return P, Q.T, e


# ### - Regularized CP tensor factorization (Multiverse Recommendation, RecSys 10)

# In[61]:

def Regularized_CP_TF(X, A, B, C, R, steps=1000, alpha=0.0002, beta=0.02):
    e = 0
    for step in range(steps):
        nNnz = 0
        # Stochastic Gradient Descent(SGD) part
        for i in range(len(X)):
            for j in range(len(X[i])):
                for k in range(len(X[i][j])):
                    if X[i][j][k] > 0:
                        eijk = X[i][j][k] - np.dot(A[i,:],np.multiply(B[j,:],C[k,:]))
                        nNnz+=1
                        for r in range(R):
                            A[i][r] = A[i][r] + alpha * (2 * eijk * B[j][r] * C[k][r] - beta * A[i][r])
                            B[j][r] = B[j][r] + alpha * (2 * eijk * A[i][r] * C[k][r] - beta * B[j][r])
                            C[k][r] = C[k][r] + alpha * (2 * eijk * B[j][r] * A[i][r] - beta * C[k][r])                            
        
        # Stop condition for SGD
        e = 0
        for i in range(len(X)):
            for j in range(len(X[i])):
                for k in range(len(X[i][j])):
                    if X[i][j][k] > 0:
                        e = e + pow(X[i][j][k] - np.dot(A[i,:],np.multiply(B[j,:],C[k,:])), 2)
#                         for r in range(R):
#                             e = e + (beta/2) * (pow(A[i][r],2) + pow(B[j][r],2) + pow(C[k][r],2))
        # Convert error to RMSE
        e = math.sqrt(e/nNnz)
        if e < 0.3:
            print ("[TF for RS] Target RMSE(0.3) is reached.")
            break
        if (step%100==0) :
            print ("[TF for RS] "+str(step)+"/"+str(steps)+" steps done.")
    return A,B,C,e


# ### - Input utility matrix

# In[68]:

def Read_Utility_Matrix():
    # Read meta data to build dictionary in order to change matrix to tensor.
    metadata_file = open("full_meta_data.txt","r")
    lines = metadata_file.readlines()
    item_context_dic = {}
    nContext = 0

    # For now, context is Artist ID
    for line in lines:
        try:            
            item = int(line.split("\t")[0])
            context = int(line.split("\t")[2])
        except:
            pass
        else:
            item_context_dic[item] = context
            if (nContext < context+1) :
                nContext = context+1
            
    metadata_file.close()
    print ("[TF for RS] Reading Metadata done.")
    
    # Read matrix data
    rating_file = open("ratings.txt","r")
    lines = rating_file.readlines()
    users, items, ratings = [], [], []
    nUser, nItem = 0, 0
    
    for line in lines:
        user = int(line.split("\t")[0])
        item = int(line.split("\t")[1])
        rating = float(line.split("\t")[2])
        users.append(user)
        items.append(item)
        ratings.append(rating)
        if (nUser < user+1) :
            nUser = user+1
        if (nItem < item+1) :
            nItem = item+1
            
    rating_file.close()
#     utility_matrix = sparse.coo_matrix((ratings, (users, items)), shape=(nUser, nItem))
    X = np.zeros((nUser,nItem,nContext))
    for i in range(len(users)):
        if items[i] in item_context_dic:
            X[users[i]][items[i]][item_context_dic[items[i]]] = ratings[i]
    return X


# ### - Main function for CP

# In[69]:

if __name__ == "__main__":
#     X = [1,2,3,4,5,6,7,8,9,10,11,12]
#     X = np.array(X)
#     X = X.reshape(2,2,3)
    X = Read_Utility_Matrix()
    print ("[TF for RS] Reading ratings done.")
    R = 2

    A = np.random.rand(len(X),R)
    B = np.random.rand(len(X[0]),R)
    C = np.random.rand(len(X[0][0]),R)
    
    A, B, C, e = Regularized_CP_TF(X, A, B, C, R)
    print ("[TF for RS] RMSE : "+str(e))

#     print (A)
#     print (B)
#     print (C)
#     D = np.transpose(skt.khatrirao((C,B)))
#     print(np.matmul(A,D))
#     print(X)


# ### - Naive CP tensor factorization 

# In[53]:

# if __name__ == "__main__":
#     # Load Matlab data
#     X = Read_Utility_Matrix()
#     # Create dense tensor from numpy array
#     T = skt.dtensor(X)
#     # Decompose tensor using CP-ALS
#     P, fit, itr, exectimes = skt.cp_als(T, 2, init='random')
#     print (P.U)


# In[ ]:

File: rec_tf.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>
#include <ctype.h>

#define nUser 420000
#define nContext 1500000
#define nItem 150000
#define maxNnz 47000000
#define nTrashStr 1000
#define nRank 100
#define alpha 0.0002
#define lambda 0.02
#define epoch 10000

float randomVal()
{
    return (float)rand() / (float)RAND_MAX;
}

int main()
{

	//Time measurement
	clock_t begin;
	clock_t end;
	begin = clock();

	// Initialize variables
    FILE *meta_fp;
    int *item_context_dic;
    int item, i;
    char trashStr[nTrashStr];
    char contextStr[nTrashStr];

    // Read meta data to build dictionary in order to change matrix to tensor.
    // For now, context is Artist ID
    item_context_dic = (int*) malloc(sizeof(int)*nItem);
    meta_fp = fopen("full_meta_data.txt", "r");
    int cnt =0;
    while (fscanf(meta_fp, "%d\t%s\t%s", &item, trashStr, contextStr) >= 0) {
        int contextLen = strlen(contextStr);
        for (i=0; i<contextLen; ++i) if (!isdigit(contextStr[i])) break;
        if (i==contextLen) {
            item_context_dic[item] = atoi(contextStr);
            if (cnt%10000==0) printf("%d\t%s\n", item, contextStr);
            ++cnt;
        }
    }
    fclose(meta_fp);
    printf("[TF for RS] Item-Context dictionary stored in memory\n");
    
    FILE *rating_fp;
    int *users, *items;
    float *ratings;

    // Read rating matrix data
    rating_fp = fopen("train_data", "r");
    users = (int*) malloc(sizeof(int)*maxNnz);
    items = (int*) malloc(sizeof(int)*maxNnz);
    ratings = (float*) malloc(sizeof(float)*maxNnz);
    int nNnz = 0;
    while (fscanf(rating_fp, "%d\t%d\t%f", &users[nNnz], &items[nNnz], &ratings[nNnz]) >= 0) {
    	// printf("%d\t%d\t%f\n", users[nNnz], items[nNnz], ratings[nNnz]);
        if (ratings[nNnz] > 0) ++nNnz;
    }
    fclose(rating_fp);

    // Time Measurement to read data
    end = clock();
    printf("[TS for RS] %f sec spent to read data\n", (double)(end-begin)/CLOCKS_PER_SEC);
    begin = clock();

    
    // Regularized CP Tensor factorization with SGD optimization
    float **A, **B, **C;
    int j;
    A = (float**) malloc(sizeof(float*)*nUser);
    B = (float**) malloc(sizeof(float*)*nItem);
    C = (float**) malloc(sizeof(float*)*nContext);
    for (i=0; i<nUser; ++i) A[i] = (float*)malloc(sizeof(float)*nRank);
    for (i=0; i<nItem; ++i) B[i] = (float*)malloc(sizeof(float)*nRank);
    for (i=0; i<nContext; ++i) C[i] = (float*)malloc(sizeof(float)*nRank);        

    srand((unsigned)time(NULL));
    for (i=0; i<nUser; ++i) for (j=0; j<nRank; ++j) A[i][j] = randomVal();
    for (i=0; i<nItem; ++i) for (j=0; j<nRank; ++j) B[i][j] = randomVal();
    for (i=0; i<nContext; ++i) for (j=0; j<nRank; ++j) C[i][j] = randomVal();

    int step;
    int steps = epoch;
    int k,r,idx;
    float val, rmse_error;

    end = clock();
    printf("[TS for RS] %f sec spent to initialize factor matrices\n", (double)(end-begin)/CLOCKS_PER_SEC);
    begin = clock();


    // Check for optimization algoirhtm to converge
    int history_size = 1000;
    float rmse_history[history_size];
    int check_history_flag = 0;
    float epsilon_converge = 0.0000001;
    float target_error = 0.3;

    for (step=0; step<steps; ++step) {
        for (idx=0; idx<nNnz; ++idx) {
            i = users[idx];
            j = items[idx];
            k = item_context_dic[items[idx]];
            val = ratings[idx];

            if (val > 0) {
                float predicted_val = 0;
                for (r=0; r<nRank; ++r) predicted_val += A[i][r]*B[j][r]*C[k][r];
                float eijk = val - predicted_val;
                for (r=0; r<nRank; ++r) {
                    A[i][r] = A[i][r] + alpha * (2 * eijk * B[j][r] * C[k][r] - lambda * A[i][r]);
                    B[j][r] = B[j][r] + alpha * (2 * eijk * A[i][r] * C[k][r] - lambda * B[j][r]);
                    C[k][r] = C[k][r] + alpha * (2 * eijk * B[j][r] * A[i][r] - lambda * C[k][r]);
                }
            }
        }
        
        rmse_error = 0;
        for (idx=0; idx<nNnz; ++idx) {
            i = users[idx];
            j = items[idx];
            k = item_context_dic[items[idx]];
            val = ratings[idx];

            if (val > 0) {
                float predicted_val = 0;
                for (r=0; r<nRank; ++r) predicted_val += A[i][r]*B[j][r]*C[k][r];
                rmse_error += pow(val-predicted_val,2);
            }
        }
        rmse_error = sqrt(rmse_error/nNnz);
        if (rmse_error<target_error) {
            printf("[TF for RS] Target RMSE(%f) is reached.\n", target_error);
            break;
        } 

        // Convergence test
        rmse_history[step%history_size] = rmse_error;
        if (step%history_size == history_size-1) check_history_flag = 1;
        if (check_history_flag) {
            for (r=0; r<history_size; ++r) if (fabsf(rmse_history[r]-rmse_error) > epsilon_converge) break;
            if (r==history_size) {
                printf("[TF for RS] Optimization algorithm converges.\n");
                break;
            }
        }
        if (step%1000==0) printf("[TF for RS] %d/%d steps done.\n",step,steps);
    }

    end = clock();
    printf("[TS for RS] %f sec spent to factorize data\n", (double)(end-begin)/CLOCKS_PER_SEC);
    printf("[TF for RS] RMSE : %f\n", rmse_error);

    return 0;
}
